### PR TITLE
.github: update to Go 1.18.1

### PR DESCRIPTION
**Pull request by kevinburkesegment**
Also regenerate all of the stubs, we don't need "+build" anymore
because both Go 1.17 and 1.18 use "go:build" for build tags.

### CODE REVIEW

**Review by achille-roussel — COMMENTED** [review]
Are we dropping compatibility with 1.16 if we remove +build?

**Comment by kevinburke** [comment]
> Are we dropping compatibility with 1.16 if we remove +build?

Yes, I think so, but note the go formatter is doing this automatically. So we either need to use 1.16/1.17 to generate these files always, or upgrade. 

IMO if you care enough about perf to use this library you should be given carrots to upgrade, or interested in upgrading.

**Comment by achille-roussel** [comment]
The segmentio/encoding package depends on this one, users may have indirect dependencies on segmentio/asm, that would break compilation of their programs then. 

I'm not against making the change but I think we need to carefully vet the impact, and take proactive steps to mitigate it, for example:
- document that we are requiring 1.17+ in the README
- tag a release prior to merging this change so there is a recent stable version to rely on if upgrading to 1.17+ isn't yet possible for some users
- make an internal announcement that we expect systems to build with 1.17+ moving forward (not useful for users outside our org tho)

**Comment by mmcloughlin** [comment]
For what it's worth, I think supporting the last two Go versions is a common practice since it matches the support policy of Go itself.

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -43,12 +43,9 @@ jobs:
...
       uses: actions/setup-go@v2
       with:
-        go-version: 1.17.x
+        go-version: 1.18.x
 
-    - name: Generate
-      run: make --always-make build
-
-    - name: Git Status
+    - name: Ensure stubs are up to date
```
Comment (mmcloughlin) [inline .github/workflows/go.yml:48]: This name isn't correct: more than just stub files are generated.

File: .github/workflows/go.yml [hunk 2]
```diff
@@ -43,12 +43,9 @@ jobs:
...
       run: |
-        git diff
-        test -z "$(git status --porcelain)"
+        go install github.com/kevinburke/differ@latest
```
Comment (mmcloughlin) [inline .github/workflows/go.yml:50]: Avoid `@latest` in CI? Pin a version instead?

Comment (kevinburkesegment) [reply .github/workflows/go.yml]: Done

File: .github/workflows/go.yml (hunk 1 above)
Comment (kevinburkesegment) [reply .github/workflows/go.yml]: Done

**Comment by kevinburkesegment** [comment]
Ok - I've done everything on the list here: https://github.com/segmentio/asm/pull/76#issuecomment-1105908240

**Review by achille-roussel — APPROVED** [review]

**Pull request merged**